### PR TITLE
[WIP]: base logic

### PR DESCRIPTION
**Pull request by awinogradov**

### CODE REVIEW

File: src/components/button/button.js [hunk 1]
```diff
@@ -1,31 +1,105 @@
...
+        } else {
+            this.setMod({'focused-hard': true});
+            this.setState({'focused-hard': true});
+        }
+    }
+
+    _onFocus(e) {
+        this._focus();
+        this.props.onFocus &&
+        this.props.onFocus(e, this.state);
```
Comment (nik-kor) [inline src/components/button/button.js]: is it required property?

Comment (nik-kor) [inline src/components/button/button.js]: it is better to move this call on previous line

Comment (awinogradov) [inline src/components/button/button.js]: booooriiiing)))

**Comment by Guria** [comment]
Do not forget about focused-hard modifier

**Comment by awinogradov** [comment]
It's exists;) check it!

File: src/components/checkbox/checkbox.deps.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+({
+    shouldDeps: [
+        { mods: { type: 'button' } }
```
Comment (awinogradov) [inline src/components/checkbox/checkbox.deps.js]: @nik-kor this deps on our level doesn't helps resolve template for `checkbox_type_button`, but it should work ;-)

Comment (awinogradov) [inline src/components/checkbox/checkbox.deps.js]: The same problem in radio

File: src/components/radio/radio.js [hunk 3]
```diff
@@ -7,6 +7,8 @@ module.exports = class Radio extends Control {
 
     _onChange() {
         this.props.disabled || this.setStateAndMod({checked: true});
+
+        this.props.onChange && this.props.onChange(e, this.state);
```
Comment (nik-kor) [inline src/components/radio/radio.js]: check this decorator https://github.com/edli/react-callback-props-decorator .

File: src/components/button/button.js [hunk 4]
```diff
@@ -72,7 +72,7 @@ module.exports = class Button extends Control {
     render() {
         return provide({
             block: this.bem.block,
-            attrs: {
+            attrs: Object.assign({
                 onMouseEnter: this._onMouseEnter.bind(this),
```
Comment (nik-kor) [inline src/components/button/button.js]: maybe I missed it but is this the handlers like _onMouseEnter came from the base class? So why should we write them here?

File: src/components/button/button.js [hunk 5]
```diff
@@ -81,7 +81,7 @@ module.exports = class Button extends Control {
                 onMouseDown: this._onMouseDown.bind(this),
                 onMouseUp: this._onMouseUp.bind(this),
                 'aria-disabled': this.props.disabled
-            },
+            }, this.props.attrs),
```
Comment (nik-kor) [inline src/components/button/button.js]: I think that it's better to pass handlers like simple props, not grouping them into attrs

File: src/components/button/button.js (hunk 4 above)
Comment (awinogradov) [inline src/components/button/button.js]: We sent bindings to JXST. Base class doesn't to this. We can move common logic to Control, your are right.

File: src/components/checkbox/__control/checkbox__control.bemhtml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+block('checkbox').elem('control')(
+    tag()('input'),
+
+    attrs()(function() {
+        const originalAttrs = applyNext();
+
+        originalAttrs.autoComplete = originalAttrs.autocomplete;
```
Comment (nik-kor) [inline src/components/checkbox/__control/checkbox__control.bemhtml]: can't we pass autoComplete as an attribute? what's the point of this template btw?

File: src/components/button/button.js (hunk 5 above)
Comment (awinogradov) [inline src/components/button/button.js]: Yes!!! This is fast experiment:)

File: src/components/checkbox/__control/checkbox__control.bemhtml (hunk 6 above)
Comment (awinogradov) [inline src/components/checkbox/__control/checkbox__control.bemhtml]: https://github.com/awinogradov/bem-components/blob/v2/common.blocks/checkbox/__control/checkbox__control.bemhtml#L6 remember previous levels ;)

File: src/components/checkbox/_type/checkbox_type_button.bemhtml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        const Button = require('../src/components/button/button');
+
+        const ctx = this.ctx,
+            mods = this.mods;
+
+        const originalContent = applyNext();
+
+        // Replace XJST Button
+        // FIXME: support icon!
+        originalContent[0] = <Button
```
Comment (nik-kor) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: yeah, that's strange)

File: src/components/checkbox/__control/checkbox__control.bemhtml (hunk 6 above)
Comment (awinogradov) [inline src/components/checkbox/__control/checkbox__control.bemhtml]: Maybe we can solve this problem here https://github.com/bem-contrib/react-xjst/issues/6

File: src/components/checkbox/_type/checkbox_type_button.bemhtml [hunk 8]
```diff
@@ -0,0 +1,27 @@
+block('checkbox').mod('type', 'button')(
+    content()(function() {
+        const React = require('react');
+        const Button = require('../src/components/button/button');
+
+        const ctx = this.ctx,
+            mods = this.mods;
+
+        const originalContent = applyNext();
```
Comment (nik-kor) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: is it required that button will be at the first position?

File: src/components/checkbox/_type/checkbox_type_button.bemhtml (hunk 7 above)
Comment (awinogradov) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: You must believe) It's works!

File: src/components/checkbox/_type/checkbox_type_button.bemhtml (hunk 8 above)
Comment (awinogradov) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: This [temlate](https://github.com/awinogradov/bem-components/blob/v2/common.blocks/checkbox/_type/checkbox_type_button.bemhtml#L6) says yes

File: src/components/input/__control/input__control.bemhtml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+block('input').elem('control')(
+    tag()('input'),
+
+    attrs()(function() {
+        var input = this._input,
+            attrs = {
```
Comment (nik-kor) [inline src/components/input/__control/input__control.bemhtml]: why can't we pass it through the react component?

File: src/components/checkbox/_type/checkbox_type_button.bemhtml (hunk 8 above)
Comment (awinogradov) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: It's right if previous level is bem-components, but if you add another level before this template we can get truble

Comment (nik-kor) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: very fragile

File: src/components/input/__control/input__control.bemhtml (hunk 9 above)
Comment (awinogradov) [inline src/components/input/__control/input__control.bemhtml]: the same problem in readio and checkbox

File: src/components/checkbox/_type/checkbox_type_button.bemhtml (hunk 8 above)
Comment (awinogradov) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: you must know right levels order – you are developer ;)

File: src/components/input/__control/input__control.bemhtml (hunk 9 above)
Comment (awinogradov) [inline src/components/input/__control/input__control.bemhtml]: https://github.com/awinogradov/react-components/pull/6#discussion_r59118231

File: src/components/button/button.js [hunk 10]
```diff
@@ -1,33 +1,99 @@
...
+    }
+
+    _onBlur(e) {
+        this.setStateAndMod({pressed: false, 'focused-hard': false});
+
+        super._onBlur(e);
+    }
+
+    _onClick(e) {
+        this.props.onClick && this.onClick(e, this.state);
```
Comment (nik-kor) [inline src/components/button/button.js]: it's not good to pass inner state of component somewhere in callback. What's the point of it?

Comment (awinogradov) [inline src/components/button/button.js]: It's not necessary it's copypaste :)

File: src/components/control/control.js [hunk 11]
```diff
@@ -1,47 +1,38 @@
...
-    //     this.setMod({hovered: false});
-    //     this.props.onMouseLeave && this.props.onMouseLeave(e, this.state);
-    // };
-    //
-    // _handleBlur(e) {
-    //     this.setMod({focused: false, hovered: false});
-    //     this.props.onBlur && this.props.onBlur(e, this.state);
-    // };
+
+    constructor(props) {
```
Comment (nik-kor) [inline src/components/control/control.js]: I think that it's useless

Comment (awinogradov) [inline src/components/control/control.js]: 👍

File: src/components/checkbox/_type/checkbox_type_button.bemhtml (hunk 7 above)
Comment (awinogradov) [inline src/components/checkbox/_type/checkbox_type_button.bemhtml]: @miripiruni @tadatuta what do you think?

**Comment by awinogradov** [comment]
@nik-kor https://github.com/alfa-bank-dev/bem-loader/issues/10

**Comment by awinogradov** [comment]
@nik-kor and maybe we can do this https://facebook.github.io/react/blog/2016/04/07/react-v15.html

**Comment by nik-kor** [comment]
@awinogradov how is v15 affect us really?

**Comment by awinogradov** [comment]
Fashionably, stylishly, youth ;)

**Comment by awinogradov** [comment]
@nik-kor https://facebook.github.io/react/blog/2016/04/07/react-v15.html#document.createelement-is-in-and-data-reactid-is-out https://facebook.github.io/react/blog/2016/04/07/react-v15.html#no-more-extra-ltspangts https://facebook.github.io/react/blog/2016/04/07/react-v15.html#rendering-null-now-uses-comment-nodes can helps us make clean structrures

**Comment by nik-kor** [comment]
@awinogradov the new version won't affect our code(I believe). It will just work. 
Also, we found that it will be hard not to wrap text with spans 
https://github.com/bem/bem-xjst/pull/235#commitcomment-16968504 - I found the workaround but it was strange https://github.com/bem/bem-xjst/pull/235/commits/3e6a5c778e215c4a5dc9cf76d59977668fa7a3f9#diff-22c95dd657a561227217908259f2c2dfR18

File: src/components/dropdown/_switcher/dropdown_switcher_link.bemhtml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    const React = require('react');
+    const Link = require('../src/components/link/link');
+    const dropdown = this._dropdown;
+    const switcher = dropdown.switcher;
+
+    if(Array.isArray(switcher)) return switcher;
+
+    let dropdownMods = this.mods;
+    return this.isSimple(switcher) ?
+        applyCtx(<Link
```
Comment (awinogradov) [inline src/components/dropdown/_switcher/dropdown_switcher_link.bemhtml]: @nik-kor it doesn't work. The first attrs it object but react waits string or class. Can you check applyCtx?

File: src/components/select/select.bemhtml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+        const Popup = require('../src/components/popup/popup');
+
+        return [
+            { elem : 'button' },
+            <Popup
+                directions={['bottom-left', 'bottom-right', 'top-left', 'top-right']}
+                target='anchor',
+                theme={this.mods.theme}
+                autoclosable={true}>
+                {{ block : this.block, mods : this.mods, elem : 'menu' }}
```
Comment (awinogradov) [inline src/components/select/select.bemhtml]: @nik-kor I found very very big trouble! Any ideas what we must do if we need react component in composition and it must contain bemjson? We can check this in components and use provide for render but I'm not sure.

File: src/components/checkbox-group/checkbox-group.bemhtml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+        const React = require('react');
+        const Checkbox = require('../src/components/checkbox/checkbox');
+
+        if(isValDef && !Array.isArray(val)) throw Error('checkbox-group: val must be an array');
+
+        return (ctx.options || []).map(function(option, i) {
+            return [
+                !!i && !mods.type && { tag : 'br' },
+                <Checkbox
```
Comment (awinogradov) [inline src/components/checkbox-group/checkbox-group.bemhtml]: @nik-kor the same problem ;(

**Comment by Guria** [comment]
`button_pressed` doesn't set on keyboard press (spacebar or enter)

**Pull request merged**